File: src/main/java/com/merge/api/resources/hris/types/RemoteResponse.java
package com.merge.api.resources.hris.types;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonSetter;
import com.fasterxml.jackson.annotation.Nulls;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.merge.api.core.ObjectMappers;
import java.util.Map;
import java.util.Objects;
import java.util.Optional;

@JsonInclude(JsonInclude.Include.NON_EMPTY)
@JsonDeserialize(builder = RemoteResponse.Builder.class)
public final class RemoteResponse {
    private final String method;

    private final String path;

    private final int status;

    private final JsonNode response;

    private final Optional<Map<String, JsonNode>> responseHeaders;

    private final Optional<RemoteResponseResponseType> responseType;

    private final Optional<Map<String, JsonNode>> headers;

    private RemoteResponse(
            String method,
            String path,
            int status,
            JsonNode response,
            Optional<Map<String, JsonNode>> responseHeaders,
            Optional<RemoteResponseResponseType> responseType,
            Optional<Map<String, JsonNode>> headers) {
        this.method = method;
        this.path = path;
        this.status = status;
        this.response = response;
        this.responseHeaders = responseHeaders;
        this.responseType = responseType;
        this.headers = headers;
    }

    @JsonProperty("method")
    public String getMethod() {
        return method;
    }

    @JsonProperty("path")
    public String getPath() {
        return path;
    }

    @JsonProperty("status")
    public int getStatus() {
        return status;
    }

    @JsonProperty("response")
    public JsonNode getResponse() {
        return response;
    }

    @JsonProperty("response_headers")
    public Optional<Map<String, JsonNode>> getResponseHeaders() {
        return responseHeaders;
    }

    @JsonProperty("response_type")
    public Optional<RemoteResponseResponseType> getResponseType() {
        return responseType;
    }

    @JsonProperty("headers")
    public Optional<Map<String, JsonNode>> getHeaders() {
        return headers;
    }

    @Override
    public boolean equals(Object other) {
        if (this == other) return true;
        return other instanceof RemoteResponse && equalTo((RemoteResponse) other);
    }

    private boolean equalTo(RemoteResponse other) {
        return method.equals(other.method)
                && path.equals(other.path)
                && status == other.status
                && response.equals(other.response)
                && responseHeaders.equals(other.responseHeaders)
                && responseType.equals(other.responseType)
                && headers.equals(other.headers);
    }

    @Override
    public int hashCode() {
        return Objects.hash(
                this.method,
                this.path,
                this.status,
                this.response,
                this.responseHeaders,
                this.responseType,
                this.headers);
    }

    @Override
    public String toString() {
        return ObjectMappers.stringify(this);
    }

    public static MethodStage builder() {
        return new Builder();
    }

    public interface MethodStage {
        PathStage method(String method);

        Builder from(RemoteResponse other);
    }

    public interface PathStage {
        StatusStage path(String path);
    }

    public interface StatusStage {
        ResponseStage status(int status);
    }

    public interface ResponseStage {
        _FinalStage response(JsonNode response);
    }

    public interface _FinalStage {
        RemoteResponse build();

        _FinalStage responseHeaders(Optional<Map<String, JsonNode>> responseHeaders);

        _FinalStage responseHeaders(Map<String, JsonNode> responseHeaders);

        _FinalStage responseType(Optional<RemoteResponseResponseType> responseType);

        _FinalStage responseType(RemoteResponseResponseType responseType);

        _FinalStage headers(Optional<Map<String, JsonNode>> headers);

        _FinalStage headers(Map<String, JsonNode> headers);
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public static final class Builder implements MethodStage, PathStage, StatusStage, ResponseStage, _FinalStage {
        private String method;

        private String path;

        private int status;

        private JsonNode response;

        private Optional<Map<String, JsonNode>> headers = Optional.empty();

        private Optional<RemoteResponseResponseType> responseType = Optional.empty();

        private Optional<Map<String, JsonNode>> responseHeaders = Optional.empty();

        private Builder() {}

        @Override
        public Builder from(RemoteResponse other) {
            method(other.getMethod());
            path(other.getPath());
            status(other.getStatus());
            response(other.getResponse());
            responseHeaders(other.getResponseHeaders());
            responseType(other.getResponseType());
            headers(other.getHeaders());
            return this;
        }

        @Override
        @JsonSetter("method")
        public PathStage method(String method) {
            this.method = method;
            return this;
        }

        @Override
        @JsonSetter("path")
        public StatusStage path(String path) {
            this.path = path;
            return this;
        }

        @Override
        @JsonSetter("status")
        public ResponseStage status(int status) {
            this.status = status;
            return this;
        }

        @Override
        @JsonSetter("response")
        public _FinalStage response(JsonNode response) {
            this.response = response;
            return this;
        }

        @Override
        public _FinalStage headers(Map<String, JsonNode> headers) {
            this.headers = Optional.of(headers);
            return this;
        }

        @Override
        @JsonSetter(value = "headers", nulls = Nulls.SKIP)
        public _FinalStage headers(Optional<Map<String, JsonNode>> headers) {
            this.headers = headers;
            return this;
        }

        @Override
        public _FinalStage responseType(RemoteResponseResponseType responseType) {
            this.responseType = Optional.of(responseType);
            return this;
        }

        @Override
        @JsonSetter(value = "response_type", nulls = Nulls.SKIP)
        public _FinalStage responseType(Optional<RemoteResponseResponseType> responseType) {
            this.responseType = responseType;
            return this;
        }

        @Override
        public _FinalStage responseHeaders(Map<String, JsonNode> responseHeaders) {
            this.responseHeaders = Optional.of(responseHeaders);
            return this;
        }

        @Override
        @JsonSetter(value = "response_headers", nulls = Nulls.SKIP)
        public _FinalStage responseHeaders(Optional<Map<String, JsonNode>> responseHeaders) {
            this.responseHeaders = responseHeaders;
            return this;
        }

        @Override
        public RemoteResponse build() {
            return new RemoteResponse(method, path, status, response, responseHeaders, responseType, headers);
        }
    }
}

File: merge-java-core/src/main/kotlin/dev/merge/api/models/RemoteResponse.kt
package dev.merge.api.models

import com.fasterxml.jackson.annotation.JsonAnyGetter
import com.fasterxml.jackson.annotation.JsonAnySetter
import com.fasterxml.jackson.annotation.JsonCreator
import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import dev.merge.api.core.ExcludeMissing
import dev.merge.api.core.JsonField
import dev.merge.api.core.JsonMissing
import dev.merge.api.core.JsonValue
import dev.merge.api.core.NoAutoDetect
import dev.merge.api.core.toUnmodifiable
import dev.merge.api.errors.MergeInvalidDataException
import java.util.Objects
import java.util.Optional

/**
 * # The RemoteResponse Object
 *
 * ### Description
 *
 * The `RemoteResponse` object is used to represent information returned from a third-party
 * endpoint.
 *
 * ### Usage Example
 *
 * View the `RemoteResponse` returned from your `DataPassthrough`.
 */
@JsonDeserialize(builder = RemoteResponse.Builder::class)
@NoAutoDetect
class RemoteResponse
private constructor(
    private val method: JsonField<String>,
    private val path: JsonField<String>,
    private val status: JsonField<Long>,
    private val response: JsonValue,
    private val responseHeaders: JsonValue,
    private val responseType: JsonField<ResponseTypeEnum>,
    private val headers: JsonValue,
    private val additionalProperties: Map<String, JsonValue>,
) {

    private var validated: Boolean = false

    private var hashCode: Int = 0

    fun method(): String = method.getRequired("method")

    fun path(): String = path.getRequired("path")

    fun status(): Long = status.getRequired("status")

    /**
     * - `JSON` - JSON
     * - `BASE64_GZIP` - BASE64_GZIP
     */
    fun responseType(): Optional<ResponseTypeEnum> =
        Optional.ofNullable(responseType.getNullable("response_type"))

    @JsonProperty("method") @ExcludeMissing fun _method() = method

    @JsonProperty("path") @ExcludeMissing fun _path() = path

    @JsonProperty("status") @ExcludeMissing fun _status() = status

    @JsonProperty("response") @ExcludeMissing fun _response() = response

    @JsonProperty("response_headers") @ExcludeMissing fun _responseHeaders() = responseHeaders

    /**
     * - `JSON` - JSON
     * - `BASE64_GZIP` - BASE64_GZIP
     */
    @JsonProperty("response_type") @ExcludeMissing fun _responseType() = responseType

    @JsonProperty("headers") @ExcludeMissing fun _headers() = headers

    @JsonAnyGetter
    @ExcludeMissing
    fun _additionalProperties(): Map<String, JsonValue> = additionalProperties

    fun validate() = apply {
        if (!validated) {
            method()
            path()
            status()
            responseType()
            validated = true
        }
    }

    fun toBuilder() = Builder().from(this)

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }

        return other is RemoteResponse &&
            this.method == other.method &&
            this.path == other.path &&
            this.status == other.status &&
            this.response == other.response &&
            this.responseHeaders == other.responseHeaders &&
            this.responseType == other.responseType &&
            this.headers == other.headers &&
            this.additionalProperties == other.additionalProperties
    }

    override fun hashCode(): Int {
        if (hashCode == 0) {
            hashCode =
                Objects.hash(
                    method,
                    path,
                    status,
                    response,
                    responseHeaders,
                    responseType,
                    headers,
                    additionalProperties,
                )
        }
        return hashCode
    }

    override fun toString() =
        "RemoteResponse{method=$method, path=$path, status=$status, response=$response, responseHeaders=$responseHeaders, responseType=$responseType, headers=$headers, additionalProperties=$additionalProperties}"

    companion object {

        @JvmStatic fun builder() = Builder()
    }

    class Builder {

        private var method: JsonField<String> = JsonMissing.of()
        private var path: JsonField<String> = JsonMissing.of()
        private var status: JsonField<Long> = JsonMissing.of()
        private var response: JsonValue = JsonMissing.of()
        private var responseHeaders: JsonValue = JsonMissing.of()
        private var responseType: JsonField<ResponseTypeEnum> = JsonMissing.of()
        private var headers: JsonValue = JsonMissing.of()
        private var additionalProperties: MutableMap<String, JsonValue> = mutableMapOf()

        @JvmSynthetic
        internal fun from(remoteResponse: RemoteResponse) = apply {
            this.method = remoteResponse.method
            this.path = remoteResponse.path
            this.status = remoteResponse.status
            this.response = remoteResponse.response
            this.responseHeaders = remoteResponse.responseHeaders
            this.responseType = remoteResponse.responseType
            this.headers = remoteResponse.headers
            additionalProperties(remoteResponse.additionalProperties)
        }

        fun method(method: String) = method(JsonField.of(method))

        @JsonProperty("method")
        @ExcludeMissing
        fun method(method: JsonField<String>) = apply { this.method = method }

        fun path(path: String) = path(JsonField.of(path))

        @JsonProperty("path")
        @ExcludeMissing
        fun path(path: JsonField<String>) = apply { this.path = path }

        fun status(status: Long) = status(JsonField.of(status))

        @JsonProperty("status")
        @ExcludeMissing
        fun status(status: JsonField<Long>) = apply { this.status = status }

        @JsonProperty("response")
        @ExcludeMissing
        fun response(response: JsonValue) = apply { this.response = response }

        @JsonProperty("response_headers")
        @ExcludeMissing
        fun responseHeaders(responseHeaders: JsonValue) = apply {
            this.responseHeaders = responseHeaders
        }

        /**
         * - `JSON` - JSON
         * - `BASE64_GZIP` - BASE64_GZIP
         */
        fun responseType(responseType: ResponseTypeEnum) = responseType(JsonField.of(responseType))

        /**
         * - `JSON` - JSON
         * - `BASE64_GZIP` - BASE64_GZIP
         */
        @JsonProperty("response_type")
        @ExcludeMissing
        fun responseType(responseType: JsonField<ResponseTypeEnum>) = apply {
            this.responseType = responseType
        }

        @JsonProperty("headers")
        @ExcludeMissing
        fun headers(headers: JsonValue) = apply { this.headers = headers }

        fun additionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.clear()
            this.additionalProperties.putAll(additionalProperties)
        }

        @JsonAnySetter
        fun putAdditionalProperty(key: String, value: JsonValue) = apply {
            this.additionalProperties.put(key, value)
        }

        fun putAllAdditionalProperties(additionalProperties: Map<String, JsonValue>) = apply {
            this.additionalProperties.putAll(additionalProperties)
        }

        fun build(): RemoteResponse =
            RemoteResponse(
                method,
                path,
                status,
                response,
                responseHeaders,
                responseType,
                headers,
                additionalProperties.toUnmodifiable(),
            )
    }

    class ResponseTypeEnum
    @JsonCreator
    private constructor(
        private val value: JsonField<String>,
    ) {

        @com.fasterxml.jackson.annotation.JsonValue fun _value(): JsonField<String> = value

        override fun equals(other: Any?): Boolean {
            if (this === other) {
                return true
            }

            return other is ResponseTypeEnum && this.value == other.value
        }

        override fun hashCode() = value.hashCode()

        override fun toString() = value.toString()

        companion object {

            @JvmField val JSON = ResponseTypeEnum(JsonField.of("JSON"))

            @JvmField val BASE64_GZIP = ResponseTypeEnum(JsonField.of("BASE64_GZIP"))

            @JvmStatic fun of(value: String) = ResponseTypeEnum(JsonField.of(value))
        }

        enum class Known {
            JSON,
            BASE64_GZIP,
        }

        enum class Value {
            JSON,
            BASE64_GZIP,
            _UNKNOWN,
        }

        fun value(): Value =
            when (this) {
                JSON -> Value.JSON
                BASE64_GZIP -> Value.BASE64_GZIP
                else -> Value._UNKNOWN
            }

        fun known(): Known =
            when (this) {
                JSON -> Known.JSON
                BASE64_GZIP -> Known.BASE64_GZIP
                else -> throw MergeInvalidDataException("Unknown ResponseTypeEnum: $value")
            }

        fun asString(): String = _value().asStringOrThrow()
    }
}
